File: design_analysis_algorithms/question E file indexing btree/file_indexing_btree.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class BTree {
private:
    struct Node {
        vector<string> keys;  // File names (keys)
        vector<string> children;  // Locations (file locations)
        vector<Node*> child_nodes;  // Child nodes (pointers to child nodes)
        bool is_leaf;

        Node(bool leaf = true) : is_leaf(leaf) {}
    };

    Node* root;
    int t;  // Minimum degree (defines the range of keys per node)

    void insertNonFull(Node* node, const string& file_name, const string& location);
    void split(Node* parent, int i);
    Node* search(Node* node, const string& file_name);

public:
    BTree(int degree = 3);
    void insert(const string& file_name, const string& location);
    string search(const string& file_name);
};

BTree::BTree(int degree) : t(degree), root(new Node()) {}

void BTree::insert(const string& file_name, const string& location) {
    Node* root = this->root;
    if (root->keys.size() == (2 * t) - 1) {
        Node* new_node = new Node(false);
        new_node->child_nodes.push_back(root);
        split(new_node, 0);
        this->root = new_node;
    }
    insertNonFull(this->root, file_name, location);
}

void BTree::insertNonFull(Node* node, const string& file_name, const string& location) {
    int i = node->keys.size() - 1;
    if (node->is_leaf) {
        while (i >= 0 && file_name < node->keys[i]) {
            i--;
        }
        node->keys.insert(node->keys.begin() + i + 1, file_name);
        node->children.insert(node->children.begin() + i + 1, location);
    } else {
        while (i >= 0 && file_name < node->keys[i]) {
            i--;
        }
        i++;
        Node* child = node->child_nodes[i];
        if (child->keys.size() == (2 * t) - 1) {
            split(node, i);
            if (file_name > node->keys[i]) {
                i++;
            }
        }
        insertNonFull(node->child_nodes[i], file_name, location);
    }
}

void BTree::split(Node* parent, int i) {
    Node* node = parent->child_nodes[i];
    Node* new_node = new Node(node->is_leaf);
    int mid_index = node->keys.size() / 2;
    parent->keys.insert(parent->keys.begin() + i, node->keys[mid_index]);
    parent->child_nodes.insert(parent->child_nodes.begin() + i + 1, new_node);

    new_node->keys = vector<string>(node->keys.begin() + mid_index + 1, node->keys.end());
    node->keys.resize(mid_index);

    if (!node->is_leaf) {
        new_node->child_nodes = vector<Node*>(node->child_nodes.begin() + mid_index + 1, node->child_nodes.end());
        node->child_nodes.resize(mid_index + 1);
    }
}

BTree::Node* BTree::search(Node* node, const string& file_name) {
    int i = 0;
    while (i < node->keys.size() && file_name > node->keys[i]) {
        i++;
    }
    if (i < node->keys.size() && node->keys[i] == file_name) {
        return node;
    }
    if (node->is_leaf) {
        return nullptr;
    }
    return search(node->child_nodes[i], file_name

File: design_analysis_algorithms/question E file indexing btree/file_indexing_btree.py
class BTree:
    class Node:
        def __init__(self, is_leaf=True):
            self.is_leaf = is_leaf
            self.keys = []  # List of file names (keys)
            self.children = []  # List of children nodes

    def __init__(self, t=3):  # Default B-tree of minimum degree 3
        self.root = self.Node()
        self.t = t  # Minimum degree (defines the range of keys per node)

    def insert(self, file_name, location):
        root = self.root
        if len(root.keys) == (2 * self.t) - 1:
            # Root is full, split it
            new_node = self.Node(is_leaf=False)
            new_node.children.append(self.root)
            self._split(new_node, 0)
            self.root = new_node

        self._insert_non_full(self.root, file_name, location)

    def _insert_non_full(self, node, file_name, location):
        i = len(node.keys) - 1
        if node.is_leaf:
            # Insert the file directly into the leaf node
            while i >= 0 and file_name < node.keys[i]:
                i -= 1
            node.keys.insert(i + 1, file_name)
            node.children.insert(i + 1, location)
        else:
            # Find the child node to insert into
            while i >= 0 and file_name < node.keys[i]:
                i -= 1
            i += 1
            child = node.children[i]
            if len(child.keys) == (2 * self.t) - 1:
                self._split(node, i)
                if file_name > node.keys[i]:
                    i += 1
            self._insert_non_full(node.children[i], file_name, location)

    def _split(self, parent, i):
        node = parent.children[i]
        new_node = self.Node(is_leaf=node.is_leaf)
        mid_index = len(node.keys) // 2
        parent.keys.insert(i, node.keys[mid_index])
        parent.children.insert(i + 1, new_node)

        new_node.keys = node.keys[mid_index + 1:]
        node.keys = node.keys[:mid_index]

        if not node.is_leaf:
            new_node.children = node.children[mid_index + 1:]
            node.children = node.children[:mid_index + 1]

    def search(self, file_name):
        return self._search(self.root, file_name)

    def _search(self, node, file_name):
        i = 0
        while i < len(node.keys) and file_name > node.keys[i]:
            i += 1
        if i < len(node.keys) and node.keys[i] == file_name:
            return node.children[i]  # Return the location of the file

        if node.is_leaf:
            return None
        return self._search(node.children[i], file_name)

# Example Usage
btree = BTree(t=3)

# Insert files into the B-tree
btree.insert("file1.txt", "Location 1")
btree.insert("file2.txt", "Location 2")
btree.insert("file3.txt", "Location 3")
btree.insert("file4.txt", "Location 4")

# Search for a file by name
location = btree.search("file2.txt")
print(f"File found at: {location}")
